### PR TITLE
Fix Collection/Map serialization

### PR DESCRIPTION
**Pull request by jpbempel**
# What Does This Do
Catch exception and report it safely into the snapshot

# Motivation
Some implementation of collections or maps may throw exceptions in somce situations which can happen during our serialization process 

# Additional Notes

### CODE REVIEW

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/util/SerializerWithLimits.java [hunk 1]
```diff
@@ -166,15 +169,33 @@ public void serialize(Object value, String type, Limits limits) throws Exception
...
       tokenWriter.collectionPrologue(value);
-      Collection<?> col = (Collection<?>) value;
-      boolean isComplete = serializeCollection(col, limits);
-      tokenWriter.collectionEpilogue(value, isComplete, col.size());
+      Collection<?> col;
+      boolean isComplete = true;
+      int size = 0;
+      try {
+        col = (Collection<?>) value;
+        size = col.size(); // /!\ alien call /!\
```
Comment (ojung) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/util/SerializerWithLimits.java:177]: Where is the synchronization that makes this an alien call or do you mean something else than that? Btw, how come you're adding the comments now? Did we run into any deadlock?

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/SnapshotSerializationTest.java [hunk 2]
```diff
@@ -717,6 +718,56 @@ public void capturedValueAdapterNull() {
     Assertions.assertEquals("null", capturedValueAdapter.toJson(null));
   }
 
+  @Test
+  public void collectionValueThrows() throws IOException {
+    JsonAdapter<Snapshot> adapter = createSnapshotAdapter();
+    Snapshot snapshot = createSnapshot();
+    CapturedContext context = new CapturedContext();
+    CapturedContext.CapturedValue mapField =
```
Comment (ojung) [inline dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/SnapshotSerializationTest.java:726]: Var should probably be called `listField`.

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/SnapshotSerializationTest.java [hunk 3]
```diff
@@ -717,6 +718,56 @@ public void capturedValueAdapterNull() {
...
+            new ArrayList<String>() {
+              @Override
+              public int size() {
+                throw new UnsupportedOperationException();
+              }
+            });
+    context.addFields(new CapturedContext.CapturedValue[] {mapField});
+    snapshot.setExit(context);
+    String buffer = adapter.toJson(snapshot);
+    System.out.println(buffer);
```
Comment (ojung) [inline dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/SnapshotSerializationTest.java:739]: Left-over or intended?

**Review by ojung — APPROVED** [review]
Thanks, just a couple of questions / minor things.

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/util/SerializerWithLimits.java (hunk 1 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/util/SerializerWithLimits.java:177]: "Alien call" means we are calling an external code that we do not control and can be throwing whatever exception. see unit tests.
There is no synchronization issue per se, but I am reusing the same terminology as this is similar issue here.

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/SnapshotSerializationTest.java (hunk 3 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/SnapshotSerializationTest.java:739]: intended, for serialization tests I am keeping the println for troubleshooting purpose

File: dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/SnapshotSerializationTest.java (hunk 2 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/test/java/com/datadog/debugger/agent/SnapshotSerializationTest.java:726]: ah yes, good catch

**Review by ojung — APPROVED** [review]

**Pull request merged**